File: plugins/plugin-evm/python/elizaos_plugin_evm/actions/gov_queue.py
import logging

from elizaos_plugin_evm.error import EVMError
from elizaos_plugin_evm.providers.wallet import EVMWalletProvider
from elizaos_plugin_evm.types import QueueParams

logger = logging.getLogger(__name__)


async def execute_queue(
    provider: EVMWalletProvider,
    params: QueueParams,
) -> str:
    logger.info(
        "Queueing governance proposal on governor %s on %s",
        params.governor,
        params.chain.value,
    )

    if not params.targets or len(params.targets) == 0:
        raise EVMError.invalid_params("Targets array cannot be empty")

    from web3 import Web3

    w3 = Web3()
    queue_selector = w3.keccak(text="queue(address[],uint256[],bytes[],bytes32)")[:4]
    encoded_params = w3.codec.encode(
        ["address[]", "uint256[]", "bytes[]", "bytes32"],
        [params.targets, params.values, params.calldatas, params.description_hash],
    )

    data = queue_selector.hex() + encoded_params.hex()

    tx_hash = await provider.send_transaction(
        chain=params.chain,
        to=params.governor,
        value=0,
        data=data,
    )

    await provider.wait_for_transaction(params.chain, tx_hash)

    logger.info("Proposal queued: %s", tx_hash)
    return tx_hash


queue_action = {
    "name": "QUEUE",
    "description": "Queue a passed DAO governance proposal for execution",
    "similes": [
        "GOVERNANCE_QUEUE",
        "queue_proposal",
    ],
    "examples": [
        "Queue proposal on governor 0x1234 on Ethereum",
        "Queue the passed proposal for execution on Base",
    ],
    "handler": execute_queue,
}

File: plugins/plugin-evm/typescript/actions/gov-queue.ts
import type { ActionResult, HandlerCallback, IAgentRuntime, Memory, State } from "@elizaos/core";
import { type Address, encodeFunctionData, type Hex, keccak256, stringToHex } from "viem";
import governorArtifacts from "../contracts/artifacts/OZGovernor.json";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { WalletProvider } from "../providers/wallet";
import { queueProposalTemplate } from "../templates";
import type { QueueProposalParams, SupportedChain, Transaction } from "../types";

export { queueProposalTemplate };

const spec = requireActionSpec("QUEUE_PROPOSAL");

export class QueueAction {
  constructor(private walletProvider: WalletProvider) {}

  async queue(params: QueueProposalParams): Promise<Transaction> {
    const walletClient = this.walletProvider.getWalletClient(params.chain);

    if (!walletClient.account) {
      throw new Error("Wallet account is not available");
    }

    const descriptionHash = keccak256(stringToHex(params.description));

    const txData = encodeFunctionData({
      abi: governorArtifacts.abi,
      functionName: "queue",
      args: [params.targets, params.values, params.calldatas, descriptionHash],
    });

    try {
      const chainConfig = this.walletProvider.getChainConfigs(params.chain);
      const publicClient = this.walletProvider.getPublicClient(params.chain);

      // @ts-expect-error - viem type narrowing issue with sendTransaction parameters
      const hash = await walletClient.sendTransaction({
        account: walletClient.account,
        to: params.governor,
        value: BigInt(0),
        data: txData as Hex,
        chain: chainConfig,
      });

      const receipt = await publicClient.waitForTransactionReceipt({
        hash,
      });

      return {
        hash,
        from: walletClient.account?.address as `0x${string}`,
        to: params.governor,
        value: BigInt(0),
        data: txData as Hex,
        chainId: this.walletProvider.getChainConfigs(params.chain).id,
        logs: receipt.logs,
      };
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      throw new Error(`Queue failed: ${errorMessage}`);
    }
  }
}

export const queueAction = {
  name: spec.name,
  description: spec.description,
  handler: async (
    runtime: IAgentRuntime,
    _message: Memory,
    _state: State,
    options: Record<string, unknown>,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    try {
      if (
        !options.chain ||
        !options.governor ||
        !options.targets ||
        !options.values ||
        !options.calldatas ||
        !options.description
      ) {
        throw new Error("Missing required parameters for queue proposal");
      }

      const queueParams: QueueProposalParams = {
        chain: options.chain as SupportedChain,
        governor: options.governor as Address,
        targets: options.targets as Address[],
        values: (options.values as string[]).map((v) => BigInt(v)),
        calldatas: options.calldatas as `0x${string}`[],
        description: String(options.description),
      };

      const privateKey = runtime.getSetting("EVM_PRIVATE_KEY") as `0x${string}`;
      const walletProvider = new WalletProvider(privateKey, runtime);
      const action = new QueueAction(walletProvider);
      const result = await action.queue(queueParams);
      return {
        success: true,
        text: `Proposal queued successfully. Transaction hash: ${result.hash}`,
        data: {
          transactionHash: result.hash,
          from: result.from,
          to: result.to,
          chain: queueParams.chain,
        },
      };
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      if (callback) {
        callback({ text: `Error: ${errorMessage}` });
      }
      return {
        success: false,
        text: `Error: ${errorMessage}`,
      };
    }
  },
  template: queueProposalTemplate,
  validate: async (runtime: IAgentRuntime) => {
    const privateKey = runtime.getSetting("EVM_PRIVATE_KEY");
    return typeof privateKey === "string" && privateKey.startsWith("0x");
  },
  examples: [
    [
      {
        name: "user",
        content: {
          text: "Queue proposal 123 on the governor at 0x1234567890123456789012345678901234567890 on Ethereum",
          action: "QUEUE_PROPOSAL",
        },
      },
    ],
    [
      {
        name: "user",
        content: {
          text: "Queue the passed proposal for execution on governor 0xabcdef1234567890abcdef1234567890abcdef12 on Base",
          action: "QUEUE_PROPOSAL",
        },
      },
    ],
    [
      {
        name: "user",
        content: {
          text: "Add proposal 789 to the timelock queue on the DAO at 0x9876543210987654321098765432109876543210 on Arbitrum",
          action: "QUEUE_PROPOSAL",
        },
      },
    ],
  ],
  similes: spec.similes ? [...spec.similes] : [],
};
